Create specification and generate tests for kernels based on hand written labels

reikdas opened the pull request:
Fixes #300 
Fixes #301
Fixes #302

This is what the result looks like - 
![image](https://user-images.githubusercontent.com/11775615/85202299-0c4fb900-b323-11ea-8501-3e3790ef1335.png)

@jpivarski Is this how we want it? I don't know if we should be enforcing a particular labelling method via `//` in the .cpp files (I updated CONTRIBUTING.md to reflect this) - but I don't see a way around it.

(I'll create a different PR for the input parameter roles - they will be documented as Doxygen's `@param`)

reikdas committed:
Define compiler intrinsic

reikdas committed:
Revert "Define compiler intrinsic"

This reverts commit 4bf93b2b456a98094cd75bf7539831729179fa28.

reikdas committed:
Demonstrate demo interface

jpivarski reviewed (commented):
On the names, why not "input" and "output"?

This syntax is not doxygen-compatible: doxygen requires three slashes `///` or in a slash-double-star block `/**` (see doxygen's documentation). Then, everything for a parameter follows a directive like

```c++
/// @param name whatever
```

This makes the bulleted list of parameters on the doxygen HTML with "name (type) whatever" on each line. We can use the "whatever" for the "input"/"output" and (if input) the "role" specifications.

If you were planning on putting only the roles in doxygen, you'd have to have two separate lists of comments above each function, and that would get unwieldy.

reikdas committed:
Merge branch 'master' into reikdas/label

reikdas committed:
Change interface based on PR review

Add role detection and Python docstring generation

reikdas committed:
Working demo of ctypes

reikdas committed:
Create entire functions for all function type variants

Might revert

reikdas committed:
Changed handling of bool ISOPTION template param

reikdas committed:
Fix in hardcoded sorting.cpp python

reikdas committed:
Revert "Create entire functions for all function type variants"

This reverts only part of 314f2cc9536434ad44b57e6ed30103e93bca7477 for a
better diff.

reikdas committed:
Fix hardcoded Python for sorting.cpp

reikdas committed:
x

reikdas committed:
Revert "x"

This reverts commit 3e4789d49f2de1ee5e5f0736ee04daf67450d2b6.

reikdas committed:
Checkpoint before testing hypothesis

reikdas committed:
Update CONTRIBUTING.md

reikdas committed:
Put doxygen style comment blocks only in header files

reikdas committed:
Remove recursion in kernels

reikdas committed:
Tests working for a specific kernel

reikdas committed:
Change all dict to OrderedDict

Probably not all dicts need to be OrderedDict, but it is good to
defensive program this to avoid difficult to debug errors.

reikdas committed:
Update documentation

reikdas committed:
1 test per kernel works for identities.cpp

reikdas committed:
Fixed broken kernel test in identities

jpivarski reviewed (commented):
I see that most of the first "role" definitions are in include/awkward/cpu-kernels/identities.h. You and Anish both start with this file most often—maybe because it's shortest? The Identities subsystem is, unfortunately, one of the least tested, so I think it's not a good place to start, but I wholly support the idea of developing the ideas on something small before scaling up.

I think you're going to run into a problem with the roles you've defined not being specific enough. I mean you'll run into a problem because when you try making tests, you'll want more information from the function to make appropriate arrays. (dev/testcases.json is just a start.)

As a starter set, I think you're going to want

   * NumpyArray-ptr
   * NumpyArray-shape
   * NumpyArray-strides
   * NumpyArray-offset
   * RegularArray-size
   * ListArray-starts
   * ListArray-starts-offset
   * ListArray-stops
   * ListArray-stops-offset
   * ListOffsetArray-offsets
   * ListOffsetArray-offsets-offset
   * IndexedArray-index
   * IndexedArray-index-offset
   * IndexedOptionArray-index
   * IndexedOptionArray-index-offset
   * ByteMaskedArray-mask
   * ByteMaskedArray-mask-offset
   * ByteMaskedArray-valid_when
   * BitMaskedArray-mask
   * BitMaskedArray-mask-offset
   * BitMaskedArray-mask-length
   * BitMaskedArray-valid_when
   * BitMaskedArray-lsb_order
   * UnionArray-tags
   * UnionArray-tags-offset
   * UnionArray-index
   * UnionArray-index-offset

as well as a role instance (i.e. `0`, `1`, maybe even `2`), so that the arguments that belong to the same ListArray can be drawn from the same package of samples but arguments coming from two different ListArrays should be Cartesian-producted.

Then when you generate test examples, you can define a set like

```python
{"samples":
 {"class": "ListArray",
  "samples": [
    {"ListArray-starts": [0, 3, 3, 5],
     "ListArray-stops":  [3, 3, 5, 10],
     "ListArray-starts-offset": 0,
     "ListArray-stops-offset": 0},
    {"ListArray-starts": [99, 99, 99, 0, 3, 3, 5],
     "ListArray-stops":  [99, 99, 99, 3, 3, 5, 10],
     "ListArray-starts-offset": 3,
     "ListArray-stops-offset": 3},
    ...
  ]},
 {"class": "ListOffsetArray",
  ...
```

A kernel with a set of roles and role instances like

```
arg0: ListArray-starts 0
arg1: ListArray-stops 0
arg2: ListArray-starts-offset 0
arg3: ListArray-stops-offset 0
arg4: ListArray-starts 1
arg5: ListArray-stops 1
arg6: ListArray-starts-offset 1
arg7: ListArray-stops-offset 1
```

would get tested at 4 sets of input values for the 2 samples defined above (2×2). That is, we bring in starts, stops, starts-offset, and stops-offset from the same group where they have the same role instance, from a different group where they have a different role instance, so the number of times a function gets run is the product of the number of samples defined for each distinct (role, role instance) pair.

I didn't include `Index` lengths in the starter list of roles above because it's not interesting to test: the kernels cannot determine if they're walking off the end of an array—all they know is the length they're given. It _could_ be interesting to test `length` parameter values that are shorter than the true length, which is legal input, but I don't see how that would catch any bugs, whether the kernels are implemented in CPU or GPU. I did include all the `offset` parameters, correlated with array values, because this will likely catch bugs. Several `offset` related bugs have already been found—they're not well tested because `offset` is often, but not always, zero. However, correlating them in a group, so that the right `offset` is used with each array, is crucial.

These sets of roles do not attempt to correlate e.g. a ListArray's `Indexes` (`starts` and `stops`) with its `content`. Thus, if a kernel takes a `contentlen` (a role that I didn't include above; it's one of the roles that is not associated with a class), then this `contentlen` is not correlated with the `starts` and `stops` and some combinations should raise error conditions. That's good: we want to see that. (We don't want every test to be testing an error condition, but a reasonable balance.) The ListArray samples should include cases with both small and big `starts` and `stops` values, and the `contentlen` samples should include both small and big values. About half of the Cartesian product would be error cases and half would be non-error cases.

In case you're thinking that only an exact match between the largest `starts` and `stops` and `contentlen` is valid, I'd like to dispel that misconception: [all of the validity rules defined here](https://awkward-array.readthedocs.io/en/latest/ak.layout.Content.html) are inequalities: e.g. the `content` of a ListArray has to be big enough to satisfy the `starts` and `stops`, but there's no limit on how big it can be. (Most examples in practice are perfectly matched, but the validity rules are as permissive as possible, which simplifies test-generation.)

Oh, and the `random` method on all the classes defined in the above link generate valid combinations of parameters. That can be a starting point for generating your test cases.

reikdas committed:
Detect functions returning failure

reikdas committed:
Fix displaying kernel roles as docstring in documentation

reikdas committed:
Redefine more specific roles for tests

Start off with operations

reikdas committed:
recurse studies

reikdas committed:
testing..

reikdas committed:
Some cpp functions don't have corresponding header

reikdas committed:
Merge branch 'master' of https://github.com/scikit-hep/awkward-1.0 into reikdas/label

reikdas committed:
Format parser_utils with black

reikdas committed:
Improve tokenizer

reikdas committed:
Working example from new roles

reikdas committed:
Add support for testing more functions in operations.cpp

and also re-design the test json.

reikdas committed:
Revert "testing.."

This reverts commit ffa2317b3d8afe9ae7b0cc3c17aa19f66df12eea.

reikdas committed:
Revert "recurse studies"

This reverts commit 89d9a0f22088b82ed5e562c2c33d914ab23656f1.

reikdas committed:
Revert "Remove recursion in kernels"

This reverts commit 229b3f5e6ef1b41cc3bd1bc9f9452c1291603729.

reikdas committed:
Convert template casts to float instead of int

to add support for more kernels in operations

reikdas committed:
Add support for remaining functions in operations

reikdas committed:
Fix failing function parser

reikdas committed:
Add support for testing failure

also print testing information

reikdas committed:
Do not hard code path to libawkward-cpu-kernels.so

reikdas committed:
Fix Python codegen via more explicit C++ code

reikdas committed:
Add testing infrastructure for getitem kernels

reikdas committed:
Add the new testing infrastructure for identities kernels

reikdas committed:
Merge branch 'master' of https://github.com/scikit-hep/awkward-1.0 into reikdas/label

reikdas committed:
Add testing infrastructure for reducers kernels

reikdas committed:
Remove redundant parameters in sorting kernels

reikdas committed:
Add testing infrastructure for sorting kernels

reikdas committed:
Fix test cases for existing kernels

reikdas committed:
Fix hardcoded Python translation of sorting kernel

Also fix indentation in Python translation of sorting kernel

reikdas committed:
Merge branch 'master' of https://github.com/scikit-hep/awkward-1.0 into reikdas/label

reikdas committed:
Start working on generating a specification

starting the whole refactor

reikdas committed:
Different way of finding kernel arg types

reikdas committed:
Extract Python kernels from specification

reikdas committed:
New kernel documentation generator

reikdas committed:
Update kernel doc generation requirements

reikdas committed:
Merge branch 'master' of https://github.com/scikit-hep/awkward-1.0 into reikdas/label

reikdas committed:
Format generated Python in spec with black

No need to do it separately for docgen now

reikdas committed:
Remove cast in range

Replacing template cast with float and using it as range is wrong -
float cannot be in range.

reikdas committed:
Format specgen.py with black

reikdas committed:
Replace old parameter parser with new EBNF grammar parser

Move FuncBody and FuncDecl class back inside genpython because they
don't have any use outside of it anymore.

reikdas committed:
Reorganize some functions

reikdas committed:
Ignore functions that don't return struct Error

reikdas committed:
Add input and output tests to specification

reikdas committed:
Run kernel tests based on specification

And fix some existing CPU kernels

reikdas committed:
Fixed single failing test

reikdas committed:
Cleanup

reikdas committed:
Merge branch 'master' of https://github.com/scikit-hep/awkward-1.0 into reikdas/label

reikdas committed:
Reorganize files for testing

reikdas committed:
Fixes...

reikdas committed:
Some cleanup

Actually fixes the failing test

reikdas committed:
Update CONTRIBUTING.md

reikdas committed:
Raise error if no tests for kernel

reikdas committed:
Fix dynamic loading of cpu-kernels so

reikdas committed:
Test all combination of tests in testcases.json

reikdas committed:
Revert "Update CONTRIBUTING.md"

This reverts commit 165122742887e88be0f82144fa5922a4683de74a.

reikdas committed:
Create different spec file for each kernel

reikdas committed:
Cleaning up current test generation relics

reikdas committed:
New testing method

reikdas committed:
Do not need to test failing Python kernel in genspec

and fix testing failure in Python kernels

reikdas committed:
New test generation technique - fixed combinatorics

reikdas committed:
Split cpu kernel tests into separate files

reikdas committed:
Fix retrieval of cpu-kernels.so

reikdas committed:
Tune the generated tests

reikdas committed:
Reorganize file structure

Copying parser_utils for now. Will need to fix

reikdas committed:
Merge branch 'master' of https://github.com/scikit-hep/awkward-1.0 into reikdas/label

reikdas committed:
Generate specification before generating documentation

Fix once we check spec into git

reikdas committed:
Add kernel tests to CI

Remove kernel specification generator when we check spec into git
Optional import of black for CI

reikdas committed:
New way of creating protosamples

reikdas committed:
Fix finding cpu-kernels object file

reikdas committed:
Fix some kernel tests

reikdas committed:
Fix for Python2.7 and Python3.5

and replace reload with exec for generated functions
PyYAML returns unordered dict from spec - Fix getting the order
math.isclose not available in Python 2

reikdas committed:
Make arg name uniform between kernel and specialization

reikdas committed:
Fix Windows CI

jpivarski commented on tests-kernels/__init__.py:
```diff
@@ -0,0 +1,38 @@
+            break
+
+if CPU_KERNEL_SO is None:
+    CPU_KERNEL_SO = pkg_resources.resource_filename("awkward1", name)
```
If the second method (`pkg_resources.resource_filename`) always works, it's better than searching from the `TOP_DIR`. The active `awkward1` Python package would usually be located in a different directory than the tests; this is true in CI because they're built with `pip install .` rather than `python localbuild.py`. So if you can remove everything except for the last line and it still works, that's the better way to go.

reikdas replied:
But then I think, someone who built it using `python localbuild.py` instead of with `pip` would not be able to run tests?

jpivarski replied:
I'd expect `pkg_resources.resource_filename("awkward1", name)` to point to the version in the nearby `awkward1` directory in that case. If it doesn't work and the explicit search is needed, it's still probably better to check `pkg_resources` first, since you don't want a nearby directory to overshadow the directory corresponding to `import awkward1`.

jpivarski reviewed (approved):
(It was a strange page-reloading problem.)

Since I'll be talking about files that aren't in the repo (generated kernels and tests), I'll do it inline in comments, anyway.

reikdas committed:
Different way of handling skipiped tests

reikdas committed:
Rename output variable

reikdas committed:
Use pytest approx instead of manually comparing
